### DISCUSSION TITLE
using git to sync encrypted journals

### DISCUSSION THREAD

**Discussion by thetayloredman** [jrnl-org/jrnl#1446] [Q&A]
### Environment

  - Jrnl `--diagnostic` output: 
```
jrnl: v2.8.4
Python: 3.9.5 (default, Nov 18 2021, 16:00:48) 
[GCC 10.3.0]
OS: Linux 5.11.0-49-generic
```
  - Install method:  **pipx**

### What are you trying to do?

I am trying to synchronize journals using Git. The problem here is **my jrnls must be encrypted.**

### What have you tried?

I am trying to write a merge engine to decrypt and re-encrypt the journals. The problem is I only want to prompt for a password once. I also need to read the list of journals programmatically. I am not sure how to pass the --decrypt a password from the command line (not STDIN)

### Other Information

Suggested: ENV.JRLNCRYPT_PASSWORD and `jrnl --list --style=json`

**Comment by wren** [comment]
@thetayloredman Hi, and thanks for filing an issue!

The problem that you're running into is that `--decrypt` and `--encrypt` operate on the journal file, but it seems to me like you're trying to view the *output* from jrnl (not necessarily the file). Luckily, git has a config value made just for cases like this.

The way to do this is:
1. Update your git config by running the following line in your terminal (assuming you're working with the `default` journal)
    ```bash
    git config diff.jrnl.textconv "jrnl --format text --config-override journals.default.journal"
    ```
2. Add the following line to the `.gitattributes` file in your repo:
    ```gitattributes
    *.jrnl diff=jrnl
    ```
3. Make sure that the password for this journal is stored in your keyring (otherwise git will get confused when asked for the password)

And that should be it. I just tested this locally, and I can see the diff in plain text for my test journal even though the journal file itself stays encrypted the entire time.

**Reply by thetayloredman** [reply]
Ahah! smart!

**Reply by thetayloredman** [reply]
What about merging..?

**Reply by wren** [reply]
I generally don't recommend merging encrypted files (they should basically be treated as binary). Instead, though, you could do something like the example at the end of [this page](https://git-scm.com/book/en/v2/Customizing-Git-Git-Attributes) in the git docs, and define a merge strategy using the `.gitattributes` file.

**Reply by thetayloredman** [reply]
@wren That was the point of making this issue. I want to merge journals from multiple sources with data-at-rest encryption.

**Reply by wren** [reply]
Oh, sorry about that. I thought the point was about git, but it seems that I misunderstood.

In that case, you could use [`--import`](https://jrnl.sh/en/stable/reference-command-line/#-import). You can import the same journal into itself and the entries will be de-duplicated (e.g. `jrnl --co journals.default.journal ./some_file.jrnl --format text | jrnl --import`), so you'll be left with just the different entries.

I'm not as familiar with git merge strategies, but I think you can basically use the same trick from above (with `--config-override`/`--co` and `--format text`) to have jrnl operate on files without having to decrypt anything yourself. This should allow you to do anything you'd normally do with jrnl through git instead.

Does that answer your question?

**Comment by thetayloredman** [comment]
I think so... I'll have to export both to a tmpfs however or it's not D-A-R.

**Reply by wren** [reply]
Ok, sounds good. Please let us know if you have any more questions!
